plugins/parsers/registry.go

```diff
@@ -6,24 +6,25 @@ package parsers
 import (
 	"github.com/aws/amazon-cloudwatch-agent/plugins/parsers/awscsm"
 	"github.com/aws/amazon-cloudwatch-agent/plugins/parsers/emf"
+	"github.com/influxdata/telegraf"
 	"github.com/influxdata/telegraf/plugins/parsers"
 )
 
-func NewEMFParser(config *parsers.Config) (parsers.Parser, error) {
+func EMFParserCreator(metricName string) telegraf.Parser {
+	// Telegraf will create parser with metricName only, and initFromConfig will be called to get the tags
 	return &emf.EMFParser{
-		MetricName:  config.MetricName,
-		DefaultTags: config.DefaultTags,
-	}, nil
+		MetricName: metricName,
+	}
 }
 
-func NewCSMParser(config *parsers.Config) (parsers.Parser, error) {
+func CSMParserCreator(metricName string) telegraf.Parser {
 	parser := &awscsm.JSONParser{
 		MetricName: "awscsm",
 	}
-	return parser, nil
+	return parser
 }
 
 func init() {
-	parsers.RegisterParser("emf", NewEMFParser)
-	parsers.RegisterParser("aws_csm", NewCSMParser)
+	parsers.Add("emf", EMFParserCreator)
+	parsers.Add("aws_csm", CSMParserCreator)
```

Comment:
CWA has a lot of code that names a constructor for struct "Foo" as `NewFoo()`.
I recommend following the convention and keeping the old names:
`NewEMFParser()` and `NewCSMParser()`.

Reply:
Because the method in Parsers 
`func Add(name string, creator Creator) {
`
is taking a Creator as second argument, and the method EMFParserCreator, CSMParserCreator is implementing the Creator interface, https://github.com/aws/telegraf/blob/aws-cwa-release-1-23-0/plugins/parsers/registry.go#L26, so I feel it's more appropriate to be named with xxxParserCreator, while newXXXParser is fine by me as well, I was leaning towards the Creator naming just to reflecting it's implementing that interface.
Let me know if that make sense to you please.